xmlua/c14n.lua

```diff
@@ -0,0 +1,237 @@
+local C14n = {}
+
+local libxml2 = require("xmlua.libxml2")
+local ffi = require("ffi")
+
+local DEFAULT_MODE = "C14N_EXCLUSIVE_1_0"
+
+local Attribute
+local AttributeDeclaration
+local CDATASection
+local Comment
+local Document
+local DocumentFragment
+local DocumentType
+local Element
+local ElementDeclaration
+local EntityDeclaration
+local NamespaceDeclaration
+local Notation
+local ProcessingInstruction
+local Text
+
+function C14n.lazy_load()
+  Attribute = require("xmlua.attribute")
+  AttributeDeclaration = require("xmlua.attribute-declaration")
+  Text = require("xmlua.text")
+  CDATASection = require("xmlua.cdata-section")
+  Comment = require("xmlua.comment")
+  Document = require("xmlua.document")
+  DocumentFragment = require("xmlua.document-fragment")
+  DocumentType = require("xmlua.document-type")
+  Element = require("xmlua.element")
+  ElementDeclaration = require("xmlua.element-declaration")
+  EntityDeclaration = require("xmlua.entity-declaration")
+  NamespaceDeclaration = require("xmlua.namespace-declaration")
+  Notation = require("xmlua.notation")
+  ProcessingInstruction = require("xmlua.processing-instruction")
+end
+
+
+
+local C14N_MODES = {
+  C14N_1_0            = ffi.C.XML_C14N_1_0,           -- Original C14N 1.0 spec
+  C14N_EXCLUSIVE_1_0  = ffi.C.XML_C14N_EXCLUSIVE_1_0, -- Exclusive C14N 1.0 spec
+  C14N_1_1            = ffi.C.XML_C14N_1_1,           -- C14N 1.1 spec
+}
+
+
+
+local C14N_MODES_LOOKUP = {} -- lookup by name or number, returns the number
+for name, number in pairs(C14N_MODES) do
+  C14N_MODES_LOOKUP[name] = number
+  C14N_MODES_LOOKUP[number] = number
+end
+
+
+
+-- list can be a string (space separated), an array of strings, or nil
+local function getNamespacePrefixArray(list)
+  local list = list or {}
+
+  if type(list) == "string" then
+    -- list is a string, assume it is the space separated PrefixList attribute, split it
+    local list_str = list
+    list = {}
+    list_str:gsub("([^%s]+)", function(cap) list[#list+1] = cap end)
+  end
+
+  if #list == 0 then
+    return nil
+  end
+
+  local result = ffi.new('xmlChar*[?]', #list+1)
+  local refs = {}
+  for i, prefix in ipairs(list) do
+    local c_ns = ffi.new("unsigned char[?]", #prefix+1, prefix)
+    ffi.copy(c_ns, prefix)
+    result[i-1] = c_ns
+    refs[i] = c_ns -- hold on to refs to prevent GC while in use
+  end
+  result[#list] = nil
+
+  return ffi.gc(result, function(ptr)
+    refs = nil -- release references, so they can be GC'ed
+  end)
+end
+
+
+
+local function getNodesList(nodes)
+  if (not nodes) or #nodes == 0 then
+    return nil
+  end
+
+  local nodeTab = ffi.new("xmlNodePtr[?]", #nodes)
+  for i = 1, #nodes do
+      nodeTab[i - 1] = nodes[i] -- FFI side is 0 indexed
+  end
+
+  local set = ffi.new("xmlNodeSet")
+  set.nodeNr = #nodes
+  set.nodeMax = #nodes
+  set.nodeTab = nodeTab
+
+  return set
+end
+
+
+
+local wrap_raw_node do
```

Comment:
Wow. I didn't know this syntax. What does this syntax mean?

Reply:
This syntax:

```lua
local wider_scope_func do

  local private_counter = 0
  
  local function private_func()
    -- do stuff...
  end

  function wider_scope_func(...) -- note: no local declaration here
    -- do stuff
  end
end
```

The function created is `wider_scope_func`. It uses several upvalues, `private_counter` and `private_func` in this case. By enclosing everything in a `do ... end` block, the visibility of those private upvalues is limited to only the function `wider_scope_var`. So this is safer in general. Variables cannot accidentally be reused or overwritten by code further down the module.

One could also do;
```lua
local wider_scope_var
do
  -- other stuff here
end
```

but imho starting the `do end` block on the same line as the `local` better shows the intent of the code; creating a single value.

Reply:
Thanks. I understand.